examples/android/OGVideoProcDroid/src/ogles_gpgpu/examples/ogvideoprocdroid/CPUImgProcThread.java

```diff
@@ -0,0 +1,187 @@
+    /**
+      * Output data from ogles_gpgpu library. Pixel data as ARGB bytes values.
+      */
+    private ByteBuffer imgData = null;
+
+    /**
+     * 'imgData' access synchronization.
+     */
+    private final Object lock = new Object();
+
+    /**
+     * Data availability flag. Variable is set in 'update' call and mainly
+     * needed to check that wakeup is not spurious.
+     */
+    private boolean dataReady = false;
+
+    /**
+      * Current state of the thread.
+      */
+    private boolean running = true;
+
+    /**
+      * Output histogram.
+      */
+    private float[] outputHist = new float[256];
+
+    /**
+     * Histogram view.
+     */
+    private HistView histView = null;
+
+    /**
+     * Activity that have started this thread.
+     */
+    private Activity mainActivity;
+
+    public CPUImgProcThread(HistView view, Activity activity) {
+        histView = view;
+        mainActivity = activity;
+    }
+
+    /**
+     * Thread run loop. Will run until terminate() is called.
+     */
+    @Override
+    public void run() {
+        Log.i(CamActivity.TAG, "starting thread CPUImgProcThread");
+
+        try {
+            mainLoop();
+        }
+        catch (Exception e) {
+            e.printStackTrace();
+            Log.e(CamActivity.TAG, "CPUImgProcThread mainLoop");
+        }
+
+        imgData = null;
+
+        Log.i(CamActivity.TAG, "stopped thread CPUImgProcThread");
+    }
+
+    private void mainLoop() throws Exception {
+        while (running) {
+            synchronized (lock) {
+                lock.wait();
+                if (!dataReady) {
+                    continue;
+                }
+                // calculate the histogram from the image data
+                calcHist(imgData);
+
+                // send it to histView, where a copy will be made
+                histView.getHistCopy(outputHist);
+
+                // invalidate the histView, so it will be redrawn.
+                // this *must* be called on the UI thread
+                mainActivity.runOnUiThread(new Runnable() {
+                    @Override
+                    public void run() {
+                        histView.invalidate();
+                    }
+                });
+                imgData = null;
+                dataReady = false;
+            }
+        }
+    }
+
+    /**
+     * Stop the thread.
+     */
+    public void terminate() {
+        synchronized (lock) {
+            running = false;
+            lock.notify();
+        }
+    }
+
+    /**
+     * Notify about new image data from ogles_gpgpu. The "imgData" ByteBuffer is
+     * only a reference to the actual image data on the native side! It is
+     * only valid until the next call to "getOutputPixels()"!
+     */
+    public void update(OGJNIWrapper ogWrapper) {
+        if (!running) {
+            return;
+        }
+        if (dataReady) {
+            return; // TODO; Should be `if (!tryLock) { return; }`
```

Comment:
I've tested [ReentrantLock](http://developer.android.com/reference/java/util/concurrent/locks/ReentrantLock.html) here but it seems that `synchronized (lock)` doesn't playing well with `lock.tryLock()`.

I thought that we can do:
```java
if(!lock.tryLock()) {
  return; // exit immediately
}
lock.notify();
lock.unlock();
```

But in fact it should be:
```java
if(!lock.tryLock()) {
  return;
}
synchronized (lock) {
  lock.notify();
}
lock.unlock();
```

which makes all the synchronization code looks quite ugly. I think we need to add new class or something (?)